packages/kiota_serialization_json/lib/src/json_parse_node.dart

```diff
@@ -0,0 +1,233 @@
+  }
+
+  @override
+  T? getObjectValue<T extends Parsable>(ParsableFactory<T> factory) {
+    final item = factory(this);
+    if (item is UntypedNode) {
+      return getUntypedValue(_node)! as T;
+    }
+
+    onBeforeAssignFieldValues?.call(item);
+    _assignFieldValues(item);
+    onAfterAssignFieldValues?.call(item);
+
+    return item;
+  }
+
+  void _assignFieldValues<T extends Parsable>(T item) {
+    if (_node is Map) {
+      onBeforeAssignFieldValues?.call(item);
+
+      final itemAdditionalData = item is AdditionalDataHolder
+          ? (item as AdditionalDataHolder).additionalData
+          : null;
+
+      final fieldDeserializers = item.getFieldDeserializers();
+
+      for (final entry in _node.entries) {
+        if (fieldDeserializers.containsKey(entry.key)) {
+          final fieldDeserializer = fieldDeserializers[entry.key];
+          if (fieldDeserializer != null) {
+            final itemNode = JsonParseNode(entry.value)
+              ..onBeforeAssignFieldValues = onBeforeAssignFieldValues
+              ..onAfterAssignFieldValues = onAfterAssignFieldValues;
+            fieldDeserializer.call(itemNode);
+          }
+        } else {
+          itemAdditionalData?[entry.key as String] = entry.value;
+        }
+      }
+      onAfterAssignFieldValues?.call(item);
+    }
+  }
+
+  @override
+  String? getStringValue() {
+    final result = _node.toString();
+    return result == 'null' ? null : result;
+  }
+
+  @override
+  TimeOnly? getTimeOnlyValue() {
+    return _node == null ? null : TimeOnly.fromDateTimeString(_node.toString());
+  }
+
+  UntypedNode? getUntypedValue(dynamic node) {
+    if (node == null) {
+      return UntypedNull();
+    }
+
+    if (node is List) {
+      final arrayValue = getCollectionOfUntypedValues(node);
+      return UntypedArray(arrayValue);
+    }
+
+    if (node is Map) {
+      final propertiesMap = <String, UntypedNode>{};
+      for (final entry in node.entries) {
+        final fieldKey = entry.key as String;
+        final fieldValue = entry.value;
+        final childNode = JsonParseNode(fieldValue);
+        // childNode.setOnBeforeAssignFieldValues(this.getOnBeforeAssignFieldValues());
+        // childNode.setOnAfterAssignFieldValues(this.getOnAfterAssignFieldValues());
+        propertiesMap[fieldKey] =
+            childNode.getUntypedValue(fieldValue)!;
+      }
+      return UntypedObject(propertiesMap);
+    }
+
+    final doubleValue = double.tryParse(node.toString());
+    if (doubleValue != null) {
+      return UntypedDouble(doubleValue);
+    }
+
+    final intValue = int.tryParse(node.toString());
+    if (intValue != null) {
+      return UntypedInteger(intValue);
+    }
+
+    final boolValue = bool.tryParse(node.toString());
+    if (boolValue != null) {
+      return UntypedBoolean(value: boolValue);
+    }
+
+    if (node is String) {
+      return UntypedString(node);
+    }
+
+    throw ArgumentError.value(node, 'node');
```

Comment:
nit: Can we add a descriptive message here as well?

Reply:
Done by Joanne